asset-transfer-secured-agreement/application-gateway-typescript/src/app.ts

```diff
@@ -0,0 +1,427 @@
+
+        // Org2 has not set a sale price and this should fail
+        await readSalePrice(assetKey, mspIdOrg2, contractOrg2);
+
+        // Org1 has not agreed to buy so this should fail
+        await readBidPrice(assetKey, mspIdOrg1, contractOrg1);
+
+        // Org2 should be able to see the price it has agreed
+        await readBidPrice(assetKey, mspIdOrg2, contractOrg2);
+
+        // Org1 will try to transfer the asset to Org2
+        // This will fail due to the sell price and the bid price are not the same
+        await transferAsset(assetKey, mspIdOrg1,mspIdOrg2, 110,contractOrg1);
+
+        // Agree to a sell by Org1,the seller will agree to the bid price of Org2
+        await agreeToSell(assetKey,mspIdOrg1,100,contractOrg1);
+
+        // Read the public details by  org1.
+        await readAsset(assetKey, mspIdOrg1, contractOrg1,mspIdOrg1);
+
+        // Read the public details by  org2.
+        await readAsset(assetKey, mspIdOrg1, contractOrg2,mspIdOrg2);
+
+        // Org1 should be able to read the private data details of the asset.
+        await readPrivateAsset(assetKey,mspIdOrg1,contractOrg1)
+
+        // Org1 should be able to read the sale price of this asset
+        await readSalePrice(assetKey, mspIdOrg1, contractOrg1);
+
+        // Org2 should be able to see the price it has agreed
+        await readBidPrice(assetKey, mspIdOrg2, contractOrg2);
+
+        // Org2 user will try to transfer the asset to Org1
+        // This will fail as the owner is Org1
+        await transferAsset(assetKey, mspIdOrg2,mspIdOrg2, 100,contractOrg2);
+
+        // Org1 will transfer the asset to Org2
+        // This will now complete as the sell price and the bid price are the same
+        await transferAsset(assetKey, mspIdOrg1,mspIdOrg2, 100, contractOrg1);
+
+        // Read the public details by  org1.
+        await readAsset(assetKey, mspIdOrg2, contractOrg1,mspIdOrg1);
+
+        // Read the public details by  org2.
+        await readAsset(assetKey, mspIdOrg2, contractOrg2,mspIdOrg2);
+
+        // Org2 should be able to read the private data details of this asset
+        await readPrivateAssetAfterTransfer(assetKey, mspIdOrg2, contractOrg2);
+
+        // Org1 should not be able to read the private data details of this asset,expected to fail
+        await readPrivateAssetAfterTransfer(assetKey, mspIdOrg1, contractOrg1);
+
+        // This is an update to the public state and requires only the owner to endorse.
+        // Org2 wants to indicate that the items is no longer for sale
+        await changePublicDescriptionAfterTransfer(assetKey,mspIdOrg2,`Asset ${assetKey} owned by ${mspIdOrg2} is NOT for sale`,contractOrg2);
+
+        // Read the public details by  org1.
+        await readAsset(assetKey, mspIdOrg2, contractOrg1,mspIdOrg1);
+
+        // Read the public details by  org2.
+        await readAsset(assetKey, mspIdOrg2, contractOrg2,mspIdOrg2);
+
+    } finally {
+        gatewayOrg1.close();
```

Comment:
can the `close()` itself fail?  Might be an idea to wrap each in a `try {}` such that even if one fails, the the others might still be able to clean up properly

Reply:
The TypeScript implementation of `close()` currently does nothing. It might do something in the future but we don't expect `close()` to fail